crates/rolldown_binding/src/types/binding_outputs.rs

```diff
@@ -32,45 +38,55 @@ impl BindingOutputs {
   pub fn assets(&mut self) -> Vec<BindingOutputAsset> {
     let mut assets: Vec<BindingOutputAsset> = vec![];
 
-    self.inner.iter_mut().for_each(|o| match o {
-      rolldown_common::Output::Asset(asset) => {
-        assets.push(BindingOutputAsset::new(unsafe { std::mem::transmute(asset.as_mut()) }));
-      }
-      rolldown_common::Output::Chunk(_) => {}
+    self.inner.with_inner(|inner| {
+      let mut inner = inner.lock().unwrap();
+      inner.iter_mut().for_each(|o| match o {
+        rolldown_common::Output::Asset(asset) => {
+          assets.push(BindingOutputAsset::new(unsafe { std::mem::transmute(asset.as_mut()) }));
+        }
+        rolldown_common::Output::Chunk(_) => {}
+      });
     });
     assets
   }
 
   #[napi]
   pub fn delete(&mut self, file_name: String) {
-    if let Some(index) = self.inner.iter().position(|o| o.filename() == file_name) {
-      self.inner.remove(index);
-    }
+    self.inner.with_inner(|inner| {
+      let mut inner = inner.lock().unwrap();
+      if let Some(index) = inner.iter().position(|o| o.filename() == file_name) {
+        inner.remove(index);
+      }
+    });
   }
 }
 
 /// The `FinalBindingOutputs` is used at `write()` or `generate()`, it is similar to `BindingOutputs`, if using `BindingOutputs` has unexpected behavior.
 /// TODO find a way to export it gracefully.
 #[napi]
 pub struct FinalBindingOutputs {
-  inner: Vec<rolldown_common::Output>,
+  inner: UniqueArcMutex<Vec<rolldown_common::Output>>,
```

Comment:
I think this is kind of unnecessary due to `FinalBindingOutputs` actually get the ownership of  `rolldown_common::Output`.

But so far `FinalBindingOutputs` rely on `BindingOutputChunk` data types, we need to define another `FinalBindingOutputChunk` to remove the `WeakRefMutex` from fields.

Reply:
Can't. The old `BindingOutputs` is holding a `WeakRefMutex<Output>` and that's ok, because it's caller it's holding `UniqueArcMutex<Vec<rolldown_common::Output>>`.

But for `FinalBindingOutputs`, there aren't callers. It's responsible to keep the liveness of `UniqueArcMutex<Vec<rolldown_common::Output>>` on its own.

Reply:
The `FinalBindingOutputs` is added before becuase the `BindingOutputs` has some unpexcted behavior, now it could be expected.

Reply:
As I said in https://github.com/rolldown/rolldown/pull/1905#discussion_r1709347907. It can't due to we need to ensure the liveness of `UniqueArcMutex<Vec<rolldown_common::Output>>` when there are `WeakRefMutex<Output>` try to access it.

Reply:
Ok. I get the original issue.